### ISSUE TITLE
Proposal: let content scripts listen for event to detect when the extension is unloaded (for various reasons)

### ISSUE THREAD

**Issue by bershanskiy** [hober/web-extensions#138]
## Problem
Context scripts continue running even after their extension is uninstalled or disabled, and they have no convenient way to be notified of these events to start their cleanup. I contribute to multiple extensions which need this functionality to prevent old context scripts from interfering with new context scripts after extension update.

## Current workarounds

Right now, an extension content script can detect extension uninstall/shutdown by:
 - registering a persistent port and waiting for its shutdown (port can be closed by either crash, extension uninstall or periodic shutdown)
 - try to send a message from content script to background and get error that receiving end (background context) does not exist
 - listen for `chrome.runtime.onUpdateAvailable` event (extension listens for this event, asks content scripts to clean themselves up, then installs update)

Workarounds 1 and 3 can be combined into one to achieve very precise notification timings if needed.

## Proposal
Let content pages listen for a new event, `chrome.runtime.onUninstalled`. If we want to be very helpful, we can even supply a `reason` attribute, which can be `disabled` (extension was disabled), `uninstalled` (user completely uninstalled extension), or `update` (old version was uninstalled and a new version was installed in its place).

## Response to security concerns
This problem was raised multiple times in the past and received responses like "notifying content scripts about extension shutdown is dangerous because the content script can retaliate against uninstalling and do something bad". I don't understand this argument, since clearly continuing to run an orphaned content script might be even worse.

**Comment by tophf** [comment]
Another workaround that doesn't depend on timing and seems reliable in my practice is to send a synchronous DOM message from the newly injected content script to the old orphaned one so it performs the cleanup, and then the new content script initializes. It's used in [Stylus extension](https://github.com/openstyles/stylus/blob/master/content/apply.js) (search for `orphan`, there's just a few lines).

**Comment by bershanskiy** [comment]
> Another workaround that doesn't depend on timing and seems reliable in my practice is to send a synchronous DOM message from the newly injected content script to the old orphaned one so it performs the cleanup, and then the new content script initializes. It's used in [Stylus extension](https://github.com/openstyles/stylus/blob/master/content/apply.js) (search for orphan, there's just a few lines).

Yes, that is yet another workaround. I believe some closed-source extensions like Grammarly use it too.

**Comment by bershanskiy** [comment]
Relevant Chromium bug: [414213](https://bugs.chromium.org/p/chromium/issues/detail?id=414213) created in 2014.

**Comment by rdcronin** [comment]
We discussed this at the WECG Meet-Up.

We agree this makes sense from a functionality point of view and are interested in exploring it more.  The next steps here will be to write a concrete proposal, following the new steps outlined in the [proposal process](https://github.com/w3c/webextensions/blob/main/proposals/proposal_process.md).

One note we highlighted during discussion is that we probably want to call this something other than `runtime.onUninstalled`, since the extension will likely want to know about other cases in which the extension is no longer active (disable, reload, etc).

**Comment by fregante** [comment]
> we probably want to call this something other than `runtime.onUninstalled`

Linking a related proposal, the name might be `onDisabled`:

- https://github.com/w3c/webextensions/issues/353

**Comment by rdcronin** [comment]
@carlosjeurissen volunteered to drive a proposal for this.  If that's approved, Chromium will sponsor the engineering work.

**Comment by carlosjeurissen** [comment]
Thanks @rdcronin!

@Rob--W can you share the Firefox stand?

**Comment by Rob--W** [comment]
Neutral to supportive here, but I will put supportive here because it is an occasional need that is difficult to solve reliably with alternatives.

I recognize the use cases here. When persistent background pages existed, extensions used port.onDisconnect as a hack to detect content script context invalidation, but that does not work in Firefox because it does not only invalidate the execution context, but it also nukes the execution environment. For more details and use cases, see the comments on https://bugzilla.mozilla.org/show_bug.cgi?id=1223425

The typical response to extension context invalidation across at least Chrome and Firefox (and Safari according to comments at #138) is to invalidate extension APIs. It's interesting what that would imply for the proposed API. For example, whether to invalidate after event dispatch, before, etc. If invalidated after event dispatch, my question is what should happen when the event loop is spun (e.g. by sync XHR) - should the extension APIs somehow stay alive for that indefinite amount of time? More generally, for how long is the content script context guaranteed to be available? There may be separate considerations in API invalidation (described before) and content script execution environment lifetime. Possible options include after returning from the handler, a fixed deadline, when the event listener's returned Promise settles.

**Comment by fregante** [comment]
> how long is the content script context guaranteed to be available?

As far as I know content scripts will continue to work until the document is unloaded and I don’t think that should change. The disabling, update, uninstall of extensions also shouldn't be delayed just because of this either. 

The way I've been using this even is to clean up any elements and listeners from the page to possibly prepare for a new injection (in case of reload, update).

In Chrome the options page is also left behind and the storage APIs stop working. In that case the only safe option is to reload the page (for simple pages) or tell the user that further changes will be lost, and to prompt a manual reload.
